app/lib/hooks/useOffsetParams.ts

```diff
@@ -0,0 +1,78 @@
+import { useEffect, useState } from "react";
+import { useSearchParams } from "react-router-dom";
+import {
+  InputToken,
+  inputTokens,
+  RetirementToken,
+  retirementTokens,
+} from "@klimadao/lib/constants";
+
+const inputParams = [
+  "quantity",
+  "inputToken",
+  "retirementToken",
+  "beneficiary",
+  "beneficiaryAddress",
+  "message",
+  "projectToken",
+] as const;
+
+interface OffsetParams {
+  quantity?: string;
+  inputToken?: InputToken;
+  retirementToken?: RetirementToken;
+  beneficiary?: string;
+  beneficiaryAddress?: string;
+  message?: string;
+  projectTokens?: string[];
+}
+
+/** Type guard */
+const isValidInputToken = (str: string | undefined): str is InputToken => {
+  if (typeof str === "undefined") return true;
+  return !!inputTokens.includes(str.toLocaleLowerCase() as InputToken);
+};
+/** Type guard */
+const isValidRetirementToken = (
+  str: string | undefined
+): str is RetirementToken => {
+  if (typeof str === "undefined") return true;
+  return !!retirementTokens.includes(str.toLowerCase() as RetirementToken);
+};
```

Comment:
You are right I really screwed this up, brainded late afternoon code here 😊 

1. initially I was just checking if the array includes values so that `ts` can infer the correct type.
but then I mindlessly refactored to support `undefined` params. `undefined` should be handled outside the type guard
1. I also should handle lower case conversion outside of this.
1. !! is not needed, did that out of habit from `array.find` and `array.findIndex`
Will fix, thanks!

The goal of the type guards is just to help TS along because `string` is not assignable to `RetirementToken`, but downstream expects `RetirementToken`. And also to validate that the given param value is correct